bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/teaser/CifTeaserImpl.java

```diff
@@ -0,0 +1,185 @@
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = CifTeaser.class, resourceType = CifTeaserImpl.RESOURCE_TYPE)
+public class CifTeaserImpl implements CifTeaser {
+
+    protected static final String RESOURCE_TYPE = "core/wcm/components/commerce/teaser/v1/teaser";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CifTeaserImpl.class);
+    private final List<String> hiddenImageResourceProperties = new ArrayList<String>() {
+        {
+            add(JcrConstants.JCR_TITLE);
+            add(JcrConstants.JCR_DESCRIPTION);
+        }
+    };
+    private Page productPage;
+    private Page categoryPage;
+    private String title;
+    private String linkURL;
+    private boolean actionsEnabled = false;
+    private boolean titleHidden = false;
+    private boolean descriptionHidden = false;
+    private boolean titleFromPage = false;
+    private boolean descriptionFromPage = false;
+    private List<ListItem> actions = new ArrayList<>();
+    @Inject
+    private Resource resource;
+
+    @Inject
+    private Page currentPage;
+
+    @ScriptVariable
+    private PageManager pageManager;
+
+    @Self
+    private SlingHttpServletRequest request;
+
+    private Page targetPage;
+
+    @PostConstruct
+    private void initModel() {
+        ValueMap properties = resource.getValueMap();
+        actionsEnabled = properties.get(CifTeaser.PN_ACTIONS_ENABLED, actionsEnabled);
+
+        productPage = Utils.getProductPage(currentPage);
+        categoryPage = Utils.getCategoryPage(currentPage);
+
+        titleFromPage = properties.get(CifTeaser.PN_TITLE_FROM_PAGE, titleFromPage);
+        descriptionFromPage = properties.get(CifTeaser.PN_DESCRIPTION_FROM_PAGE, descriptionFromPage);
+        linkURL = properties.get(ImageResource.PN_LINK_URL, String.class);
+
+        if (actionsEnabled) {
+            hiddenImageResourceProperties.add(ImageResource.PN_LINK_URL);
+            linkURL = null;
+            populateActions();
+            if (actions.size() > 0) {
+                ListItem firstAction = actions.get(0);
+                if (firstAction != null) {
+                    targetPage = pageManager.getPage(firstAction.getPath());
+                }
+            }
+        } else {
+            targetPage = pageManager.getPage(linkURL);
+        }
+    }
+
+    void testHelper(Resource mockedResource, Page productPage, Page categoryPage) {
+        this.resource = mockedResource;
+        this.productPage = productPage;
+        this.categoryPage = categoryPage;
+
+        this.populateActions();
+    }
+
+    private void populateActions() {
+        Resource actionsNode = resource.getChild(CifTeaser.NN_ACTIONS);
+        if (actionsNode != null) {
+            for (Resource action : actionsNode.getChildren()) {
+                actions.add(new ListItem() {
```

Comment:
this is how it was done in original sling model, (`https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/TeaserImpl.java#L189`) , should modify it to separate out a class, pl suggest.

Reply:
I let you decide what you like better. I prefer having a separate class, so it can be tested outside of the context of the sling model, which makes testing easier.
Please make sure that in your unit tests you cover all cases of the class, especially `getURL`.

Reply:
`getURL` is actually tested in `CifTeaserImplTest`.
I will separate `ListItem` implementation in a class.

Reply:
new class created `com.adobe.cq.commerce.core.components.internal.models.v1.teaser.CommerceTeaserActionItem`